### DISCUSSION TITLE
Why does pyodide re-build pure python packages such as typing-extensions?

### DISCUSSION THREAD

**Discussion by maartenbreddels** [pyodide/pyodide#4667] [Q&A] [answered]
Hi,

I see that the pure Python package typing-extensions is added here: https://github.com/pyodide/pyodide/blob/main/packages/typing-extensions/meta.yaml

What advantage does this have? Since pyodide can also install files from pypi directly, this seems to put a limit/version constraint on this package.

Regards,

Maarten

**Comment by hoodmane** [comment]
We don't actually rebuild it, we just grab the wheel from pypi. We need it because it's a transitive dependency of one or more native extensions we do include.

**Reply by maartenbreddels** [reply]
But then why is this file in the repo and do you republish it on the jsdelivr cdn. Couldn't pyodide just grab the dependency from pypi (as happens with micropip)?
Is it maybe the case that the installation happens from javascript (not micropip), and that this makes things much simpler?

I'm trying to understand the design choices here to better understand pyodide.

**Reply by hoodmane** [reply, marked as answer]
> But then why is this file in the repo

We want to lock all of the transitive dependencies of the packages we include so that what we test matches what we ship. Ideally that means they will all work together correctly.

> Couldn't pyodide just grab the dependency from pypi (as happens with micropip)?

Yes, our lockfile works with PyPI urls so that would work fine. We try to remove tests from wheels that include them and do various other things to shrink the size of some wheels. We can also insert patches or add files. But some don't get modified.

**Reply by maartenbreddels** [reply]
So, i'm free to remove for instance `typing-extensions` from `js.pyodide._api.repodata_packages` and have micropip install it from pypi? I'm not doing anything crazy right?

In fact, I want to remove any non-build dependency from this repodata_packages  structure to pick up anything modern from pypi.

The underlying reason for this is that I need a more modern version for installing OpenAI.

**Reply by hoodmane** [reply]
> So, i'm free to remove for instance typing-extensions from js.pyodide._api.repodata_packages and have micropip install it from pypi? I'm not doing anything crazy right?

Dunno, if it works it works. Does it not work to do `micropip.install("typing_extensions >= min_version_needed")`?

**Reply by maartenbreddels** [reply]
Yes, but then I run into https://github.com/pyodide/micropip/issues/103
